File: Polars.FSharp/UdfUtils.fs
namespace Polars.FSharp

open System
open Apache.Arrow
open Polars.NET.Core.Arrow 
open Polars.NET.Core.Data 

module Udf =

    // ==========================================
    // Normal Map (T -> U)
    // ==========================================
    let map (f: 'T -> 'U) : Func<IArrowArray, IArrowArray> =
        Func<IArrowArray, IArrowArray>(fun inputArray ->
            let len = inputArray.Length

            let rawGetter = ArrowReader.CreateAccessor(inputArray, typeof<'T>)

            let buffer = ColumnBufferFactory.Create(typeof<'U>, len)

            let tIn = typeof<'T>
            let isValueType = tIn.IsValueType && isNull (Nullable.GetUnderlyingType tIn)

            for i in 0 .. len - 1 do
                if inputArray.IsNull i && isValueType then
                    buffer.Add null
                else
                    let rawVal = rawGetter.Invoke i
                    let inputVal = unbox<'T> rawVal 
                    
                    let outputVal = f inputVal
                    
                    buffer.Add outputVal

            buffer.BuildArray()
        )

    // ==========================================
    // 2. Option Map (T option -> U option)
    // ==========================================
    let mapOption (f: 'T option -> 'U option) : Func<IArrowArray, IArrowArray> =
        Func<IArrowArray, IArrowArray>(fun inputArray ->
            let len = inputArray.Length
            
            // 1. Reader
            let rawGetter = ArrowReader.CreateAccessor(inputArray, typeof<'T>)
            
            // 2. Writer 
            let buffer = ColumnBufferFactory.Create(typeof<'U>, len)

            for i in 0 .. len - 1 do
                let rawVal = rawGetter.Invoke i
                
                let inputOpt = 
                    if isNull rawVal then None 
                    else Some (unbox<'T> rawVal)
                
                let outputOpt = f inputOpt
                
                match outputOpt with
                | Some v -> buffer.Add v
                | None -> buffer.Add null

            buffer.BuildArray()
        )

File: Polars.CSharp/UdfUtils.cs
using Apache.Arrow;
using Polars.NET.Core.Arrow; 
using Polars.NET.Core.Data;

namespace Polars.CSharp;

internal static class UdfUtils
{
    public static Func<IArrowArray, IArrowArray> Wrap<TIn, TOut>(Func<TIn, TOut> userFunc)
    {
        return inputArray =>
        {
            int length = inputArray.Length;

            // 1. Reader: ArrowReader
            var rawGetter = ArrowReader.CreateAccessor(inputArray, typeof(TIn));

            // 2. Writer: DbToArrowStream
            var buffer = ColumnBufferFactory.Create(typeof(TOut), length);

            // 3. Null 
            bool inputIsValueType = typeof(TIn).IsValueType && Nullable.GetUnderlyingType(typeof(TIn)) == null;

            for (int i = 0; i < length; i++)
            {
                // check null logic
                bool isNull = inputArray.IsNull(i);

                if (isNull && inputIsValueType)
                {
                    buffer.Add(null!); 
                }
                else
                {
                    TIn input = (TIn)rawGetter(i)!;
                    
                    TOut output = userFunc(input);
                    
                    buffer.Add(output!);
                }
            }

            return buffer.BuildArray();
        };
    }
}